### PR TITLE
Manual save

### PR DESCRIPTION
**Pull request by ashwin-nat**
## 📝 Description

Adding a manual save option as a workaround for the game's SESSION_START bug
Adding a new IPC option as adding manual save as a HTTP endpoint has very high grief potential
Migrated post race viewer to use IPC instead of stdin
Added manual save command handler to IPC task in backend
Refactored processFinalClassificationUpdate. separate function for updating data structures and for prepping final json
Removed dir creation responsibility from telemetry handler class. Now the save_to_disk module handles this
    Dir is created only when writing file to disk. So now empty dirs will not be created

---

## 📂 Type of change

- [ ] Bug fix 🐞
- [x] New feature 🚀
- [ ] Refactor 🔨
- [ ] Documentation 📚
- [x] Tests ✅
- [ ] Other: 

---

## 🧪 Backend Test Reports (required for Python/backend changes)

> ⚠️ If your PR includes backend Python changes, please fill out the sections below.

### ✅ Unit Tests

- [x] All unit tests pass
- Attach test command/output:

```
----------------------------------------------------------------------
Ran 300 tests in 14.201s

OK
```

To run the unit test suite
```bash
poetry run python tests/unit_tests.py
```

### ✅ Integration Tests

* [x] All integration tests pass
* Attach test command/output:

```
===== TEST RESULTS =====
Passed: 23/23

PASS: f1_23_sp_austria.f1pcap
PASS: f1_23_tt_silverstone.f1pcap
PASS: f1_24_mp_spectator_bahrain.f1pcap
PASS: f1_24_sp_austria.f1pcap
PASS: f1_24_sp_austria_flashback.f1pcap
PASS: f1_24_sp_austria_flashback_pit.f1pcap
PASS: f1_24_tt_silverstone.f1pcap
PASS: f1_25_movie_preview.f1pcap
PASS: f1_25_mp_fp_jeddah.f1pcap
PASS: f1_25_mp_fp_jeddah_player.f1pcap
PASS: f1_25_mp_fp_jeddah_player_2.f1pcap
PASS: f1_25_mp_lobby_only_solo.f1pcap
PASS: f1_25_mp_race_japan_spec.f1pcap
PASS: f1_25_mp_race_jeddah.f1pcap
PASS: f1_25_mp_solo_afk_fp_aus.f1pcap
PASS: f1_25_sp_aus_5lap_1stop.f1pcap
PASS: f1_25_sp_aus_5lap_1stop_flashback.f1pcap
PASS: f1_25_sp_austria_reverse.f1pcap
PASS: f1_25_sp_baku_mid_session_save_load.f1pcap
PASS: f1_25_sp_full_gp_dnf_monaco.f1pcap
PASS: f1_25_sp_imola_5lap_1stop.f1pcap
PASS: f1_25_story.f1pcap
PASS: f1_25_tt_zandvoort_rev.f1pcap
```

To run the integration test
```bash
poetry run python tests/integration_test/runner.py
```

### ✅ Pylint

* [x] Code passes `pylint`
* Paste summary report:

```
$ pylint --rcfile scripts/.pylintrc apps lib

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

To run pylint
```bash
pylint --rcfile scripts/.pylintrc lib apps
```

---

## 🧱 `lib/` Directory Changes

* [x] This PR modifies or adds files under `lib/`
* [x] I have updated or added unit tests to reflect those changes

Explain your test updates here:

```
added support for empty final classification object - added tests for this as well
added unknown visual and actual tyre compounds - added tests for this as well
added ipc module for req/rep - added test suite for this
```

---

## 📋 General Checklist

* [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) guide
* [x] My code follows project style conventions
* [x] All new and existing tests pass
* [x] I have added relevant documentation/comments
* [x] I have reviewed my changes and believe they are ready to merge

---

## Summary by Sourcery

Add manual save functionality and robust IPC framework, refactor final classification JSON construction, centralize file saving logic, migrate UI to IPC, and introduce comprehensive IPC and save-to-disk tests

New Features:
- Add manual save command via IPC with corresponding launcher and backend handlers
- Implement ZeroMQ-based IPC framework with IpcParent, IpcChildSync, and IpcChildAsync
- Introduce save_to_disk module to handle JSON file writing and directory creation

Enhancements:
- Refactor processFinalClassificationUpdate into separate state update and JSON build functions with version, overtakes, and records
- Migrate telemetry post-race viewer to IPC and enhance open_file_helper with comprehensive error handling
- Remove directory creation from telemetry handler and delegate to save_to_disk

Build:
- Add pyzmq and update cffi constraints in requirements and pyproject.toml

Tests:
- Add unit tests for IPC communication scenarios
- Add tests for saving JSON to disk and empty final classification packets for F1 2023–2025

### CODE REVIEW

File: apps/save_viewer/telemetry_post_race_data_viewer.py [hunk 1]
```diff
@@ -1068,23 +1069,41 @@ def _getTyreWearHistory(driver_data : Dict[str, Any], start_lap : int, end_lap :
...
-            should_write = False
-            should_write |= checkRecomputeJSON(g_json_data)
-        sendRaceTable()
-        if g_should_open_ui and open_webpage:
-            g_should_open_ui = False
-            webbrowser.open(f'http://localhost:{g_port_number}', new=2)
-    png_logger.info(f"Opened file: {file_path}")
+    try:
+        with open(file_path, 'r+', encoding='utf-8') as f:
+            global g_json_lock
```
[the remark this answers is not in the document]
Comment (ashwin-nat) [reply apps/save_viewer/telemetry_post_race_data_viewer.py:1074]: there is a refactor planned for this entire code soon
